src/sw.ts

```diff
@@ -0,0 +1,274 @@
+        true,
+        result.processed,
+        Date.now() - startTime
+      ));
+      await broadcastToClients(SyncMessages.syncStatus('completed'));
+    } else {
+      throw result.error || new Error('Sync failed');
+    }
+  } catch (error) {
+    console.error('Sync error:', error);
+    
+    const errorMessage = error instanceof Error ? error.message : 'Unknown error';
+    const isNetworkError = errorMessage.includes('fetch') || errorMessage.includes('network') || 
+                           errorMessage.includes('Failed to fetch') || errorMessage.includes('NetworkError');
+    const isCancelled = errorMessage.includes('cancelled');
+    
+    // Save error to database
+    await DatabaseService.setLastSyncError(errorMessage);
+    
+    // Save checkpoint if sync was interrupted (not cancelled)
+    if (!isCancelled && currentSyncCore) {
+      // The sync core should have saved its own checkpoint during progress
+      // We just need to ensure retry count is updated
+    }
+    
+    await broadcastToClients(SyncMessages.syncError(errorMessage, isNetworkError));
+    await broadcastToClients(SyncMessages.syncStatus('failed'));
+    
+    // Schedule retry for recoverable errors
+    if (isNetworkError) {
+      const retryCount = await DatabaseService.getSyncRetryCount();
+      if (retryCount < SYNC_RETRY_DELAYS.length) {
+        const delay = SYNC_RETRY_DELAYS[retryCount];
+        await DatabaseService.incrementSyncRetryCount();
+        console.log(`Scheduling sync retry in ${delay}ms (attempt ${retryCount + 1})`);
+        setTimeout(() => {
+          (self.registration as any).sync?.register('sync-bookmarks');
+        }, delay);
+      }
+    }
+  } finally {
+    syncInProgress = false;
+    currentSyncCore = null;
+  }
+}
+
+/**
+ * Handle messages from clients
+ */
+self.addEventListener('message', async (event) => {
+  const message = event.data as SyncMessage;
+  
+  switch (message.type) {
+    case 'REQUEST_SYNC':
+      if (message.immediate) {
+        // For immediate sync, try to perform directly
+        await performSync(message.fullSync);
+      } else {
+        // For background sync, register sync event
+        await (self.registration as any).sync?.register('sync-bookmarks');
+      }
+      break;
+      
+    case 'CANCEL_SYNC':
+      if (currentSyncCore) {
+        currentSyncCore.cancelSync();
+        await broadcastToClients(SyncMessages.syncStatus('cancelled'));
+      }
+      break;
+      
+    case 'REGISTER_PERIODIC_SYNC':
+      if ('periodicSync' in self.registration) {
+        try {
+          if (message.enabled) {
+            // Register periodic sync with minimum interval (browser will decide actual frequency)
+            await (self.registration as any).periodicSync.register('periodic-sync', {
+              minInterval: message.minInterval || 12 * 60 * 60 * 1000 // 12 hours default
+            });
+            console.log('Periodic sync registered');
```

Comment:
Can the service worker also use the centralized logging so that these logs show up when debug mode is enabled?